src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypeSymbolExtensions.cs

```diff
@@ -185,5 +188,85 @@ public static bool IsConstructedFromEqualTypes(this ITypeSymbol type, ITypeSymbo
                 _ => SymbolEqualityComparer.Default.Equals(type, other)
             };
         }
+
+        /// <summary>
+        /// Reconstruct a possibly-nested type with the generic parameters of another type, accounting for type nesting and generic parameters split between different nesting levels.
+        /// </summary>
+        /// <param name="instantiatedTemplateType">The generic type from which to copy type arguments</param>
+        /// <param name="unboundConstructedType">The type to recursively instantiate</param>
+        /// <param name="numOriginalTypeArgumentsSubstituted">How many type parameters from <c><paramref name="unboundConstructedType"/>.ConstructedFrom</c> that needed to be substituted to fill the generic parameter list.</param>
+        /// <param name="extraTypeArgumentsInTemplate">How many type parameters from <paramref name="instantiatedTemplateType"/>were unused.</param>
+        /// <returns>A fully constructed type based on <c><paramref name="unboundConstructedType"/>.ConstructedFrom</c> with the generic arguments from <paramref name="instantiatedTemplateType"/>.</returns>
+        public static INamedTypeSymbol ResolveUnboundConstructedTypeToConstructedType(this INamedTypeSymbol unboundConstructedType, INamedTypeSymbol instantiatedTemplateType, out int numOriginalTypeArgumentsSubstituted, out int extraTypeArgumentsInTemplate)
+        {
+            var (typeArgumentsToSubstitute, nullableAnnotationsToSubstitute) = GetArgumentsToSubstitute(instantiatedTemplateType);
```

Comment:
Can you add comments about all the tracking of nesting and substitutions throughout? I think the logic of what needs to happen is not obvious or easy to follow.

Reply:
Thank you ❤️